compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -887,6 +984,97 @@ public void visit(BLangTypeInit cIExpr) {
         resultType = types.checkType(cIExpr, actualType, expType);
     }
 
+    public void visit(BLangWaitForAllExpr waitForAllExpr) {
+        switch (expType.tag) {
+            case TypeTags.RECORD:
+                checkTypesForRecords(waitForAllExpr);
+                break;
+            case TypeTags.MAP:
+                checkTypesForMap(waitForAllExpr.keyValuePairs, ((BMapType) expType).constraint);
+                resultType = expType;
+                break;
+            case TypeTags.NONE:
+            case TypeTags.ANY:
+                checkTypesForMap(waitForAllExpr.keyValuePairs, expType);
+                BFutureType futureType = new BFutureType(TypeTags.FUTURE, symTable.anyType, null);
+                resultType = new BMapType(TypeTags.MAP, futureType, symTable.mapType.tsymbol);
+                break;
+            default:
+                dlog.error(waitForAllExpr.pos, DiagnosticCode.INVALID_LITERAL_FOR_TYPE, expType);
+                resultType = symTable.semanticError;
+                break;
+        }
+        waitForAllExpr.type = resultType;
+    }
+
+    private void checkTypesForMap(List<BLangWaitForAllExpr.BLangWaitKeyValue> keyValuePairs, BType expType) {
+        keyValuePairs.forEach(keyVal -> checkWaitKeyValExpr(keyVal, expType));
+    }
+
+    private void checkTypesForRecords(BLangWaitForAllExpr waitExpr) {
+        List<BField> lhsFields = ((BRecordType) expType).getFields();
+        List<BLangWaitForAllExpr.BLangWaitKeyValue> rhsFields = waitExpr.getKeyValuePairs();
+
+        // fields in wait collection is more or less than the fields expected by the lhs record
+        if (rhsFields.size() != lhsFields.size()) {
```

Comment:
This will not be correct for records with optional fields. Omit the optional fields. Required fields should be present